gapis/resolve/dependencygraph/dependency_graph.go

```diff
@@ -0,0 +1,173 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dependencygraph
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/gapid/core/app/benchmark"
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapis/atom"
+	"github.com/google/gapid/gapis/capture"
+	"github.com/google/gapid/gapis/database"
+	"github.com/google/gapid/gapis/gfxapi"
+)
+
+var dependencyGraphBuildCounter = benchmark.GlobalCounters.Duration("dependencyGraph.build")
+
+type DependencyGraph struct {
+	Atoms      []atom.Atom
+	Behaviours []AtomBehaviour
+	Roots      map[StateAddress]bool
+	addressMap addressMapping
+}
+
+func (g *DependencyGraph) GetStateAddressOf(key StateKey) StateAddress {
+	return g.addressMap.addressOf(key)
+}
+
+func (g *DependencyGraph) GetHierarchyStateMap() map[StateAddress]StateAddress {
+	return g.addressMap.parent
+}
+
+func (g *DependencyGraph) SetRoot(key StateKey) {
+	g.Roots[g.GetStateAddressOf(key)] = true
+}
+
+func (g *DependencyGraph) Print(ctx context.Context, b *AtomBehaviour) {
+	for _, read := range b.Reads {
+		key := g.addressMap.key[read]
+		log.I(ctx, " - read [%v]%T%+v", read, key, key)
+	}
+	for _, modify := range b.Modifies {
+		key := g.addressMap.key[modify]
+		log.I(ctx, " - modify [%v]%T%+v", modify, key, key)
+	}
+	for _, write := range b.Writes {
+		key := g.addressMap.key[write]
+		log.I(ctx, " - write [%v]%T%+v", write, key, key)
+	}
+	if b.Aborted {
+		log.I(ctx, " - aborted")
+	}
+}
+
+type StateAddress uint32
+
+const NullStateAddress = StateAddress(0)
+
+type StateKey interface {
+	Parent() StateKey
+}
+
+type AtomBehaviour struct {
+	Reads     []StateAddress // States read by an atom.
+	Modifies  []StateAddress // States read and written by an atom.
+	Writes    []StateAddress // States written by an atom.
+	Roots     []StateAddress // States labeled as root by an atom.
+	KeepAlive bool           // Force the atom to be live.
+	Aborted   bool           // Mutation of this command aborts.
+}
+
+type addressMapping struct {
+	address map[StateKey]StateAddress
+	key     map[StateAddress]StateKey
+	parent  map[StateAddress]StateAddress
+}
+
+func (m *addressMapping) addressOf(state StateKey) StateAddress {
+	if a, ok := m.address[state]; ok {
+		return a
+	}
+	address := StateAddress(len(m.address))
+	m.address[state] = address
+	m.key[address] = state
+	m.parent[address] = m.addressOf(state.Parent())
+	return address
+}
+
+func (b *AtomBehaviour) Read(g *DependencyGraph, state StateKey) {
+	if state != nil {
+		b.Reads = append(b.Reads, g.addressMap.addressOf(state))
+	}
+}
+
+func (b *AtomBehaviour) Modify(g *DependencyGraph, state StateKey) {
+	if state != nil {
+		b.Modifies = append(b.Modifies, g.addressMap.addressOf(state))
+	}
+}
+
+func (b *AtomBehaviour) Write(g *DependencyGraph, state StateKey) {
+	if state != nil {
+		b.Writes = append(b.Writes, g.addressMap.addressOf(state))
+	}
+}
+
+type HasDependencyGraphBehaviourProvider interface {
```

Comment:
I am not sure if the "Has" prefix is very go-like.
Personally, it seems fine to me.  @ben-clayton ?